### PR TITLE
KL-419/handle dapp flow

### PR DESCRIPTION
**Pull request by Boltman92**

### CODE REVIEW

File: apps/extension/src/dapp-connection-provider/index.ts [hunk 1]
```diff
@@ -36,10 +37,13 @@ export function initializeProvider({
   maxEventListeners = 100,
   shouldSetOnWindow = true
 }: InitializeProviderOptions): InpageProvider {
+  // @ts-ignore
```
Comment (unixvb) [inline apps/extension/src/dapp-connection-provider/index.ts]: why ts-ignore?

File: libs/ui/src/utils/dapp.utils.ts [hunk 2]
```diff
@@ -27,18 +28,31 @@ const createErrorMessage = (id: string) => ({
...
+  tabs.query({}).then(queryTabs => {
+    if (isDefined(origin)) {
+      const dAppTab = queryTabs.find(tab => tab.url?.includes(origin));
+      if (isDefined(dAppTab) && isDefined(dAppTab.id)) {
+        tabs.sendMessage(dAppTab.id, createDAppResponse(messageID, result));
+      }
+    } else {
+      if (queryTabs[0].id !== undefined) {
+        tabs.sendMessage(queryTabs[0].id, createDAppResponse(messageID, result));
+      }
```
Comment (unixvb) [inline libs/ui/src/utils/dapp.utils.ts]: why `origin` is optional parameter?

Comment (unixvb) [reply libs/ui/src/utils/dapp.utils.ts]: reuse the same code from `sendResponseToDAppAndClosePopup`, `sendErrorToDAppAndClosePopup` and `sendNotificationToDApp`

**Review by unixvb — DISMISSED** [review]

**Review by MuKoJla — DISMISSED** [review]

**Review by unixvb — DISMISSED** [review]

**Review by MuKoJla — DISMISSED** [review]

File: apps/extension/src/dapp-connection-provider/index.ts [hunk 3]
```diff
@@ -36,17 +38,22 @@ export function initializeProvider({
   maxEventListeners = 100,
   shouldSetOnWindow = true
 }: InitializeProviderOptions): InpageProvider {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
```
Comment (unixvb) [inline apps/extension/src/dapp-connection-provider/index.ts]: combine this rules

File: apps/extension/src/dapp-connection-provider/inpage-provider.ts [hunk 4]
```diff
@@ -219,8 +225,16 @@ export class InpageProvider extends AbstractStreamProvider {
       this._sentWarnings.enable = true;
     }
 
+    //this.anotherProvider[0]._handleDisconnect(false);
```
Comment (unixvb) [inline apps/extension/src/dapp-connection-provider/inpage-provider.ts]: no need?

File: apps/extension/background-script.ts [hunk 5]
```diff
@@ -51,194 +44,31 @@ runtime.onConnect.addListener(async port => {
...
-
-        case 'eth_call': {
-          if (data.params !== undefined) {
-            const result = await provider.call({ to: data.params[0].to, data: data.params[0].data });
-            const response = createDAppResponse(id, result);
-
-            port.postMessage(response);
-          }
+  let isPopupOpened = false;
+  const requestStack: { message: DAppMessage }[] = [];
```
Comment (unixvb) [inline apps/extension/background-script.ts]: why not just `DAppMessage[]`?

Comment (unixvb) [reply apps/extension/background-script.ts]: and maybe `dAppMessagesStack`?

File: apps/extension/src/utils/windows.utils.ts [hunk 6]
```diff
@@ -9,32 +8,10 @@ const popupCreateData: Windows.CreateCreateDataType = {
   left: 20
 };
 
-export const openDAppConnectionConfirmationPopup = async (id: string, dAppInfo: DAppInfo) =>
+export const openPopup = async (params: { [x: string]: string }, port: Runtime.Port) => {
```
Comment (unixvb) [inline apps/extension/src/utils/windows.utils.ts]: use `Record`

File: apps/extension/background-script.ts [hunk 7]
```diff
@@ -51,194 +44,31 @@ runtime.onConnect.addListener(async port => {
...
 
-        case 'eth_sign':
-        case 'personal_sign': {
-          await openSignMessagePopup(id, data.params as string[], dAppInfo);
+      return Promise.resolve();
+    });
 
-          return Promise.resolve();
-        }
+    if ((message as InformMessage).type === 'POPUP_OPEN') {
```
Comment (unixvb) [inline apps/extension/background-script.ts]: no magic string

File: apps/extension/background-script.ts [hunk 8]
```diff
@@ -51,194 +44,31 @@ runtime.onConnect.addListener(async port => {
...
 
-        default: {
-          return Promise.reject();
-        }
-      }
+      return Promise.resolve();
     }
+    isPopupOpened
+      ? requestStack.push({ message: message as DAppMessage })
+      : await handleMessage(message as DAppMessage, port);
```
Comment (unixvb) [inline apps/extension/background-script.ts]: remove assertions please

**Review by unixvb — DISMISSED** [review]

**Review by MuKoJla — DISMISSED** [review]

File: apps/extension/background-script.ts [hunk 9]
```diff
@@ -51,213 +45,61 @@ runtime.onConnect.addListener(async port => {
...
+        isPopupOpened = false;
+        await processRequestStack();
 
-        default: {
-          return Promise.reject();
-        }
+        return Promise.resolve();
       }
+      default:
+        // @ts-ignore
```
Comment (unixvb) [inline apps/extension/background-script.ts]: remove this

File: libs/ui/src/modals/screens/d-app-connection-confirmation/d-app-connection-confirmation.tsx [hunk 10]
```diff
@@ -51,18 +52,19 @@ export const DAppConnectionConfirmation: FC = () => {
...
+  useClosePopup(params.messageId, params.dAppInfo.origin);
 
   const gasBalance = getFormattedBalance(balance.data, decimals);
 
   const sendMessage = () => {
     dispatch(connectDAppAction({ dAppInfo: params.dAppInfo, accountPublicKeyHash: selectedAccountPublicKeyHash }));
-    sendNotificationToDApp('oko_accountsChanged', [selectedAccountPublicKeyHash]);
-    sendResponseToDAppAndClosePopup(params.messageId, [selectedAccountPublicKeyHash]);
+    sendNotificationToDApp('oko_accountsChanged', [selectedAccountPublicKeyHash], params.dAppInfo.origin);
+    sendResponseToDAppAndClosePopup(params.messageId, [selectedAccountPublicKeyHash], params.dAppInfo.origin);
```
Comment (unixvb) [inline libs/ui/src/modals/screens/d-app-connection-confirmation/d-app-connection-confirmation.tsx]: let's make single params order for all messages: origin, messageId, paramsToSend

File: libs/ui/src/utils/dapp.utils.ts [hunk 11]
```diff
@@ -58,10 +39,34 @@ const createDAppNotificationResponse = <T>(method: string, params: T) => ({
...
+  setTimeout(() => close(), 1000);
+};
+
+export const sendNotificationToDApp = <T>(method: string, result: T, origin: string) => {
+  const notification = createDAppNotificationResponse(method, result);
+  sendMessageToDAppTab(notification, origin);
+};
+
+export const sendMessageToBackground = () => {
+  runtime.sendMessage(undefined, { type: 'POPUP_CLOSED' });
```
Comment (unixvb) [inline libs/ui/src/utils/dapp.utils.ts]: move POPUP_CLOSED & POPUP_OPEN into libs/ui and reexport from libs/ui/src/background-script

**Review by unixvb — DISMISSED** [review]

**Review by MuKoJla — DISMISSED** [review]
wish you luck with qa testing :wink:

**Review by yarmoshevych — DISMISSED** [review]

**Pull request merged**